### ISSUE TITLE
Should Kotlin std and Kotlin reflect really be built-in?

### ISSUE THREAD

**Issue by Omico** [square/kotlinpoet#1547]
Since Kotlin 1.4.0 the Kotlin stdlib is added by default.  [Reference here](https://kotlinlang.org/docs/whatsnew14.html#dependency-on-the-standard-library-added-by-default)

But do we really need to bundle Kotlin stdlib and Kotlin reflect inside of kotlinpoet? I was facing an issue with bundled Kotlin version being too high and cannot work with Gradle at that time (Gradle also bundled a lower version of Kotlin). Please take a look at this https://github.com/Omico/Gradm/issues/21. In my situation at that time, I may not be able to exclude Kotlin. The only available option is to bundle a modified kotlinpoet inside of my project, and finally, I decide not to do it.

The above is what I have experienced, and it is outdated. I am currently working on a Kotlin DSL wrapper for kotlinpoet https://github.com/Omico/Elucidator. When I checked dependencies through the Gradle Enterprise plugin, it occurred to me that the Kotlin stdlib is built in by default. I experimented with removing the Kotlin stdlib in my project, which worked. I tried again on kotlinpoet and replaced it in my project, and it also works (tested on Gradm). It works without adding Kotlin reflect to the project because Gradle already added it. Does Kotlin reflect to have to be used? If it must be used, do we able to tell users also import Kotlin reflect instead of built-in when importing kotlinpoet?

I have not sure if I am correct or not. Looking forward to your answer.

**Comment by ZacSweers** [comment]
Kotlin reflect could be an interesting one to consider, but reflection in kotlin quickly hits walls without it.

As for the stdlib, I don't think removing that makes sense. It more or less acts as Kotlin's runtime and excluding that for the sake of gradle's poor handling of kotlin in the build classpath isn't kotlinpoet's burden to bear. Removing that dependency wouldn't help you with issues there either as gradle doesn't really care what stdlib version any plugin brings anyway. It will always impose its own version.

**Comment by Omico** [comment]
> It will always impose its own version.

It will appear in multiple Kotlin versions in my IDEA. Previously, it made IDEA report an internal error if one of the Kotlin versions was higher than the Gradle bundled Kotlin version. But I cannot reproduce at this time. Maybe it's just IDEA has some bugs on this...

What I want to say here is, maybe it would be better for users to import the Kotlin library themselves, instead of bundled by kotlinpoet.

**Comment by Egorand** [comment]
KotlinPoet doesn't bundle neither stdlib nor reflect, it only declares them as dependencies, and library users are free to force different versions of these dependencies through their build tooling. Removing these dependencies altogether (e.g. by marking them as "compileOnly") would lead to build failures unless the user manually adds those dependencies to the classpath. Given that the use case of wanting to force custom dependency versions is most likely uncommon, and it's still possible to achieve as is, I don't think we should change anything.

**Comment by Omico** [comment]
I'm reproducing this issue in my old version of the project. Then I tried to test by building a new version of kotlinpoet with dependencies excluded. The issue still exists. It shows `Module was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.8.0, expected version is 1.6.0.` This reminds me why I said earlier that this problem could only be solved by embedding a kotlinpoet on my own and why I was saying "Such another dependency hell".

So far, the conclusion is that excluding dependencies cannot solve the Kotlin metadata binary compatibility issues. 

Closing the issue.
